File: backend/app/repositories/teaching_sessions.py
from __future__ import annotations
from typing import Any
from supabase import Client
from app.db.supabase import get_supabase_client
TEACHING_TABLE = "teaching_sessions"

class TeachingSessionsRepository:
    def __init__(self, client: Client | None = None) -> None:
        self.client = client or get_supabase_client()

    async def get_by_session_id(self, session_id: str) -> dict[str, Any] | None:
        response = (
            self.client.table(TEACHING_TABLE)
            .select("*")
            .eq("session_id", session_id)
            .limit(1)
            .execute()
        )
        data = getattr(response, "data", None) or []
        return data[0] if data else None

    async def fetch_recent(self, user_id: str, limit: int = 5) -> list[dict[str, Any]]:
        response = (
            self.client.table(TEACHING_TABLE)
            .select("raw_input, extracted_topics")
            .eq("user_id", user_id)
            .order("created_at", desc=True)
            .limit(limit)
            .execute()
        )
        return response.data or []

    async def list_for_user(self, user_id: str, limit: int | None = None) -> list[dict[str, Any]]:
        query = (
            self.client.table(TEACHING_TABLE)
            .select("*")
            .eq("user_id", user_id)
            .order("created_at", desc=True)
        )
        if limit is not None:
            query = query.limit(limit)
        response = query.execute()
        return response.data or []

    async def create(self, payload: dict[str, Any]) -> dict[str, Any]:
        response = self.client.table(TEACHING_TABLE).insert(payload).execute()
        data = getattr(response, "data", None)
        if isinstance(data, list) and data:
            return data[0]
        if isinstance(data, dict) and data:
            return data

        created = await self.get_by_session_id(payload.get("session_id"))
        if not created:
            raise RuntimeError("Teaching session insertion succeeded but record could not be retrieved")
        return created

    async def update_answer(self, session_id: str, payload: dict[str, Any]) -> dict[str, Any]:
        response = (
            self.client.table(TEACHING_TABLE)
            .update(payload)
            .eq("session_id", session_id)
            .execute()
        )

        data = getattr(response, "data", None)
        if isinstance(data, list) and data:
            return data[0]
        if isinstance(data, dict) and data:
            return data

        updated = await self.get_by_session_id(session_id)
        if not updated:
            raise RuntimeError("Teaching session update succeeded but record could not be retrieved")
        return updated

File: src/lib/database/teachingSessions.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import type { TeachingSession } from '../../types';
import { tryGetSupabaseClient, isSupabaseConfigured } from './client';
import { insertReview } from './reviews';

export async function fetchRecentTeachingHistory(
  userId: string,
  limit: number = 5
): Promise<Array<Pick<TeachingSession, 'raw_input' | 'extracted_topics'>>> {
  if (!isSupabaseConfigured) return [];
  const client = tryGetSupabaseClient();
  if (!client) return [];

  const { data, error } = await (client.from('teaching_sessions') as any)
    .select('raw_input, extracted_topics')
    .eq('user_id', userId)
    .order('created_at', { ascending: false })
    .limit(limit);

  if (error) {
    console.error('[Supabase] Failed to fetch recent teaching sessions', error);
    return [];
  }

  return (data as TeachingSession[] | null)?.map((session) => ({
    raw_input: session.raw_input,
    extracted_topics: session.extracted_topics,
  })) ?? [];
}

export async function recordTeachingSession(entry: {
  userId: string;
  sessionId: string;
  inputType: TeachingSession['input_type'];
  rawInput: string;
  extractedTopics: string[];
  followUpQuestion: string;
}): Promise<void> {
  if (!isSupabaseConfigured) return;
  const client = tryGetSupabaseClient();
  if (!client) return;

  const payload = {
    user_id: entry.userId,
    session_id: entry.sessionId,
    input_type: entry.inputType,
    raw_input: entry.rawInput,
    extracted_topics: entry.extractedTopics,
    follow_up_question: entry.followUpQuestion,
  };

  const { error } = await (client.from('teaching_sessions') as any).insert([payload]);

  if (error) {
    console.error('[Supabase] Failed to record teaching session', error);
  }
}

export async function upsertTopicWithInitialReview(params: {
  userId: string;
  topicName: string;
}): Promise<void> {
  if (!isSupabaseConfigured) return;
  const client = tryGetSupabaseClient();
  if (!client) return;

  const topicsTable = client.from('topics') as any;

  const { data: existing } = await topicsTable
    .select('*')
    .eq('user_id', params.userId)
    .eq('topic_name', params.topicName)
    .maybeSingle();

  if (existing) {
    await topicsTable
      .update({
        depth: (existing.depth || 0) + 1,
        last_reviewed: new Date().toISOString(),
      })
      .eq('id', existing.id);
    return;
  }

  const { data: newTopic, error } = await topicsTable
    .insert([
      {
        user_id: params.userId,
        topic_name: params.topicName,
        depth: 1,
      },
    ])
    .select()
    .single();

  if (error || !newTopic) {
    console.error('[Supabase] Failed to create topic', error);
    return;
  }

  const scheduledDate = new Date();
  scheduledDate.setMinutes(scheduledDate.getMinutes() + 10);

  await insertReview({
    userId: params.userId,
    topicId: newTopic.id,
    scheduledDate: scheduledDate.toISOString(),
    intervalDays: 0,
  });
}

export async function updateTeachingSessionAnswer(params: {
  sessionId: string;
  answer: string;
  qualityScore: number;
}): Promise<void> {
  if (!isSupabaseConfigured) return;
  const client = tryGetSupabaseClient();
  if (!client) return;

  const { error } = await (client.from('teaching_sessions') as any)
    .update({
      user_answer: params.answer,
      quality_score: params.qualityScore,
      intelligence_gain: Math.floor(params.qualityScore / 10),
      health_change: params.qualityScore > 70 ? 3 : 1,
    })
    .eq('session_id', params.sessionId);

  if (error) {
    console.error('[Supabase] Failed to update teaching session answer', error);
  }
}

export async function fetchTeachingSessionsByUser(
  userId: string,
  limit: number = 10
): Promise<TeachingSession[]> {
  if (!isSupabaseConfigured) return [];
  const client = tryGetSupabaseClient();
  if (!client) return [];

  const { data, error } = await (client.from('teaching_sessions') as any)
    .select('*')
    .eq('user_id', userId)
    .order('created_at', { ascending: false })
    .limit(limit);

  if (error) {
    console.error('[Supabase] Failed to fetch teaching sessions', error);
    return [];
  }

  return (data as TeachingSession[] | null) ?? [];
}
